AElf.OS.Network.Grpc/GrpcNetworkServer.cs

```diff
@@ -0,0 +1,207 @@
+            IBlockService blockService)
+        {
+            _accountService = accountService;
+            _blockService = blockService;
+            _networkOptions = options.Value;
+            
+            Logger = NullLogger<GrpcNetworkServer>.Instance;
+            EventBus = NullLocalEventBus.Instance;
+            
+            _authenticatedPeers = new List<GrpcPeer>();
+        }
+        
+        public async Task StartAsync()
+        {
+            // todo inject block service
+            var p = new GrpcPeerService(this, _blockService);
+            p.EventBus = EventBus;
+            
+            p.PeerSentDisconnection += POnPeerSentDisconnection;
+            
+            _server = new Server {
+                Services = { PeerService.BindService(p) },
+                Ports = { new ServerPort(IPAddress.Any.ToString(), _networkOptions.ListeningPort, ServerCredentials.Insecure) }
+            };
+            
+            _server.Start();
+            
+            // Add the provided boot nodes
+            if (_networkOptions.BootNodes != null && _networkOptions.BootNodes.Any())
+            {
+                foreach (var btn in _networkOptions.BootNodes)
+                {
+                    await Dial(btn);
+                }
+            }
+            else
+            {
+                Logger.LogWarning("Boot nodes list is empty.");
+            }
+        }
+
+        private void POnPeerSentDisconnection(object sender, EventArgs e)
+        {
+            if (e is PeerDcEventArgs ea)
+            {
+                _authenticatedPeers.RemoveAll(p => p.RemoteEndpoint == ea.Peer);
+            }
+        }
+
+        private async Task<bool> Dial(string address)
+        {
+            try
+            {
+                Logger.LogTrace($"Attempting to reach {address}.");
+                
+                var splitAddress = address.Split(":");
+                Channel channel = new Channel(splitAddress[0], int.Parse(splitAddress[1]), ChannelCredentials.Insecure);
+                        
+                var client = new PeerService.PeerServiceClient(channel);
+                var hsk = BuildHandshake();
+                        
+                var resp = await client.ConnectAsync(hsk);
+
+                if (resp.Success != true)
+                    return false;
+
+                _authenticatedPeers.Add(new GrpcPeer(channel, client, address, resp.Port));
+                        
+                Logger.LogTrace($"Connected to {address}.");
+
+                return true;
+            }
+            catch (Exception e)
+            {
+                Logger.LogError(e, $"Error while connection to {address}.");
+                return false;
+            }
+        }
+
+        public async Task StopAsync()
+        {
+            await _server.KillAsync();
+            
+            foreach (var peer in _authenticatedPeers)
```

Comment:
maybe we do not need to tell other peer we are offline.

other node will do ping to us, it should have a logic to keep alive.

Reply:
In my mind order is:
1. Disconnect message when we can (clean disconnect, clear connection asap, can even help debugging if we had a "reason")
2. Keepalive, this will eventually detect that the peer is offline (grpc does this).

Reply:
if we need a restart work, send shutdown message to other nodes may make the shutdown time much longer..

maybe we can leave these, and if in the future we need them, we add the feature that time.